src/Framework/IBuildEngine8.cs

```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.Build.Framework
+{
+    /// <summary>
+    /// This interface extends <see cref="IBuildEngine7" /> to allow tasks know if a warning
+    /// they logged was turned into an error.
+    /// </summary>
+    public interface IBuildEngine8 : IBuildEngine7
+    {
+        public HashSet<string> WarningsAsErrors { get; }
```

Comment:
Do we actually need to make this a public API? Or to just ensure that HasLoggedErrors works as expected?

Reply:
Good point. I'll remove public on warningsaserrors

Reply:
So do we need IBuildEngine8 then?

Reply:
I may have been confused by the original question. WarningsAsErrors needs to be visible to the tasklogginghelper so it can determine if it actually logged an error when it tries to log a warning. Without IBE8 having this hashset I'm not aware of a to get that info.

Reply:
Interface members are public by default so if you remove the public keyword nothing will change, it will still be a public member on a public interface.

So I guess to answer Mihai's question we do need an `IBuildEngine8` and it does need to have the `WarningsAsErrors` member, right? Just to confirm.

Reply:
I'm thinking, should we make the return type `ICollection<string>`? It would still be backed by a HashSet<string> but we'd return a more general type. HashSet<T> feels like an implementation detail. Contains(T) would still be implemented efficiently by the HashSet<T>.

Feels like we're painting ourselves in the corner here, instead of leaving room for potential other implementations in the future shall the need arise.

Reply:
We need an IBuildEngine8 that provides _some way_ of telling tasks that what it is about to log will turn into an error. I like the idea from your other comment suggesting we turn it into a method. I'll comment my thoughts related to that there.

Also, turning this into an ICollection<string> makes sense.

Reply:
The whole interface could be private :) But that would partition the entire supertype API in a private part and a public part. The public parts keep on revving their number, and the private part can stay constant. So we'd have a private `TheTrueBuildEngineInterface` which would extend two other interfaces: `IBuildEngine{number}`, the public bit of the API and `IPrivateBuildEngine`, the private part. On the other hand this gets way less messy if we had the final abstract class to end all versioned interfaces. The abstract class can have both public and private abstract methods. You could make this PR wait for the abstract class, or go ahead with IBE8, or do the public/private split. I think the public/private split is a bit ugly, I'd rather have a public API than yet another tiny wtf in the code :)

Reply:
I'm not sure I understand the value of the private part of the public/private split interfaces? I do like the final abstract class plan, but marcpopMSFT suggested we should wait for rainersigwald before making that kind of decision, so I shelved my planned implementation of that for now.

Reply:
> I'm not sure I understand the value of the private part of the public/private split interfaces

I was assuming it somehow prevents fixing `TaskLoggingHelper.HasLoggedErrors` to return true when logging warnings-as-errors, but it turns out it's not needed. Why do tasks require a new API in addition to the existing `TaskLoggingHelper.HasLoggedErrors`? In order to allow them to inhibit logging a warning if turns out to be an error? That does not sound like safe behaviour :)

Reply:
Ah, I see, and you're right. I got so focused on the solution that I forgot about the motivating problem. That sounds reasonable to me, though—@BenVillalobos, could this interface be internal? Maybe have IBuildEngine7 extend WarningsAsErrorsBuildEngine?

Reply:
> I think the public/private split is a bit ugly, I'd rather have a public API than yet another tiny wtf in the code :)

This is where I'm at with IBuildEngine. IMO we take one more "this is the way we've done it" IBE PR and stop at IBE9 when we implement the abstract class. Otherwise why would we extend private IBE interfaces after we add the abstract class?

It sounds like it would be even messier to have a single private IBE8, all previous IBE's are public, then we make the abstract class and have no use for private IBE's.

Reply:
Mkay, took another look at the code. So TaskLoggingHelper gets an `IBuildEngine`, and that's how it interacts with the `TaskHost:IBuildEngine{N}` which knows about warnings-as-errors from its TaskLoggingContext. Beyond keeping it public, any solution I can think of is pretty messy. So public it is then :)